consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/QbftProtocolSchedule.java

```diff
@@ -0,0 +1,167 @@
+import org.hyperledger.besu.consensus.common.bft.BftExtraDataCodec;
+import org.hyperledger.besu.ethereum.core.Address;
+import org.hyperledger.besu.ethereum.core.PrivacyParameters;
+import org.hyperledger.besu.ethereum.core.Wei;
+import org.hyperledger.besu.ethereum.mainnet.BlockHeaderValidator;
+import org.hyperledger.besu.ethereum.mainnet.MainnetBlockBodyValidator;
+import org.hyperledger.besu.ethereum.mainnet.MainnetBlockImporter;
+import org.hyperledger.besu.ethereum.mainnet.MainnetProtocolSpecs;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolScheduleBuilder;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSpecAdapters;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSpecBuilder;
+
+import java.math.BigInteger;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/** Defines the protocol behaviours for a blockchain using a QBFT consensus mechanism. */
+// TODO-jf remove duplication with the BftProtocolSchedule
+public class QbftProtocolSchedule {
+
+  private static final BigInteger DEFAULT_CHAIN_ID = BigInteger.ONE;
+
+  public static ProtocolSchedule create(
+      final GenesisConfigOptions config,
+      final PrivacyParameters privacyParameters,
+      final boolean isRevertReasonEnabled,
+      final BiFunction<Integer, Boolean, BlockHeaderValidator.Builder> blockHeaderRuleset,
+      final BftExtraDataCodec bftExtraDataCodec) {
+    final Map<Long, Function<ProtocolSpecBuilder, ProtocolSpecBuilder>> specMap = new HashMap<>();
+
+    specMap.put(
+        0L,
+        builder ->
+            applyBftChanges(
+                config.getBftConfigOptions(),
+                builder,
+                config.isQuorum(),
+                blockHeaderRuleset,
+                bftExtraDataCodec,
+                Optional.of(config.getBftConfigOptions().getBlockRewardWei()),
+                config.getQbftConfigOptions().getValidatorContractAddress().isPresent()));
+
+    final Supplier<List<QbftFork>> forks = () -> config.getTransitions().getQbftForks();
+
+    forks
+        .get()
+        .forEach(
+            fork ->
+                specMap.put(
+                    fork.getForkBlock(),
+                    builder ->
+                        applyBftChanges(
+                            config.getBftConfigOptions(),
+                            builder,
+                            config.isQuorum(),
+                            blockHeaderRuleset,
+                            bftExtraDataCodec,
+                            fork.getBlockRewardWei(),
+                            fork.getValidatorSelectionMode()
+                                .filter(m -> m == VALIDATOR_SELECTION_MODE.CONTRACT)
+                                .isPresent())));
```

Comment:
Personally, I find this level of lambda nesting quite hard to read, I wonder what your feelings are on something like...
```suggestion
for (QbftFork fork : forks.get()) {
      Function<ProtocolSpecBuilder, ProtocolSpecBuilder> applyBftChanges = builder -> applyBftChanges(
              config.getBftConfigOptions(),
              builder,
              config.isQuorum(),
              blockHeaderRuleset,
              bftExtraDataCodec,
              fork.getBlockRewardWei(),
              fork.getValidatorSelectionMode()
                      .filter(m -> m == VALIDATOR_SELECTION_MODE.CONTRACT)
                      .isPresent());
      specMap.put(
            fork.getForkBlock(),
            applyBftChanges);
    }
```

Reply:
I realise it's been copied over from elsewhere and might be beneficial to keep the structure the same until the refactor mentioned in the TODO :)

Reply:
Yeah I didn't want refactor this too heavily since I was planning to remove the duplication